sklearn/utils/testing.py

```diff
@@ -375,6 +375,38 @@ def assert_raise_message(exceptions, message, function, *args, **kwargs):
                              (names, function.__name__))
 
 
+def assert_allclose_dense_sparse(x, y, err_msg=''):
+    """Assert allclose for sparse and dense data.
+
+    Both x and y need to be either sparse or dense, they
+    can't be mixed.
+
+    Parameters
+    ----------
+    x : array-like or sparse matrix
+        First array to compare.
+
+    y : array-like or sparse matrix
+        Second array to compare.
+
+    err_msg : string, default=''
+        Error message to raise.
+    """
+    if sp.sparse.issparse(x):
+        if not sp.sparse.issparse(y):
+            raise ValueError("Can only compare two sparse matrices,"
+                             " not a sparse matrix and an array.")
+        x = x.tocsr()
+        y = y.tocsr()
+        x.eliminate_zeros()
```

Comment:
ah, you're right, it needs to be sum_duplicates. 

Based on the implementation in scipy, `sum_duplicates` returns the "canonical form" of the matrix, including sorted indices, so it should be sufficient. (provided they have the same shape, I think.)